docs/Compilation/Boogie.md

```diff
@@ -0,0 +1,23 @@
+# Dafny compilation to Boogie
+
+## `{:opaque} attribute`
+
+Internally, and for verification in Boogie only, all functions are given an extra first parameter, which is the _fuel_ (represented in Boogie by unary numbers, e.g. `$LS($LS($LZ)`) is a fuel of two).
+To unroll a function, axioms ensures that a recursive call is provided with the previous fuel minus 1, so that we don't unroll undefinitely.
+
+**Normal behavior:** When verifying each method or function, Dafny will look at every `assert`. If it finds one, every function inside it will be given a fuel of at least 2 (`$LS($LS($LZ))`), meaning the verifier will be able to unroll the function's body twice for this assertion. When this assertion is proven, Dafny provides an `assume` about the result, but in this case the functions are provided a fuel of only 1 (`$LS($LZ)`), meaning it unrolls the body only once.
```

Comment:
`every function inside it will be given ...` => `every function call inside it will use a fuel argument of at least 2`